Application/ControlMsrE2/ControlMsrE2UI.c

```diff
@@ -0,0 +1,241 @@
+
+      default:
+        Buffer[Pos] = Key.UnicodeChar;
+        ++Pos;
+        Output[0] = Key.UnicodeChar;
+        gST->ConOut->OutputString (gST->ConOut, Output);
+
+        if (Pos >= Length - 1) {
+          Buffer[Pos] = 0;
+          return Pos;
+        }
+        break;
+    }
+  } while (TRUE);
+
+  return 0;
+}
+
+CHAR16
+ReadAnyKey (
+  VOID
+  )
+{
+  CHAR16 Keys[2];
+  ReadLine (Keys, 2);
+  return Keys[0];
+}
+
+BOOLEAN
+ReadYN (
+  VOID
+  )
+{
+  CHAR16 Keys[2];
+  do {
+    ReadLine (Keys, 2);
+  } while (OcStriStr (L"yn", Keys) == 0);
+
+  return Keys[0] == 'y' || Keys[0] == 'Y';
+}
+
+#define TOKENLENGTH 32
+
+EFI_STATUS
+InterpretArguments (
+  VOID
+  )
+{
+  UINTN       Argc;
+  CHAR16      **Argv;
+  CHAR16      *Parameter;
+  UINTN       ParameterCount;
+  CHAR16      *Token;
+  UINT16      TokenIndex;
+  UINT32      Index;
+
+  Flags = 0;
+
+  if (EFI_ERROR (GetArguments (&Argc, &Argv))) {
+    Argc = 1;
+  }
+
+  ParameterCount = 0;
+
+  for (Index = 1; Index < Argc; ++Index) {
+    Token = AllocatePool (StrSize (Argv[Index]));
+
+    if (Token != NULL) {
+      StrCpyS (Token, StrLen (Argv[Index]) + 1, Argv[Index]);
+
+      TokenIndex = 0;
+
+      while (Argv[Index][TokenIndex] != '\0') {
+        while (Argv[Index][TokenIndex] == ' ') {
+          ++TokenIndex;
+        }
+
+        if (Argv[Index][TokenIndex] != '\0') {
+          Parameter = &Token[TokenIndex];
+
+          while (Token[TokenIndex] != '\0' && Token[TokenIndex] != ' ') {
+            ++TokenIndex;
+          }
+          Token[TokenIndex] = '\0';
+
+          if (!StrCmp (Parameter, L"lock")) {
+            Flags |= ARG_LOCK;
+            ++ParameterCount;
+          } else if (!StrCmp (Parameter, L"unlock")) {
+            Flags |= ARG_UNLOCK;
+            ++ParameterCount;
+          } else if (!StrCmp (Parameter, L"interactive")) {
+            Flags |= ARG_INTERACTIVE;
+            ++ParameterCount;
+          } else {
+            Print (L"Ignoring unknown command line argument: %s\n", Parameter);
+          }
+        }
+      }  ///<  All Tokens parsed
+      FreePool (Token);
+    } else {
+      Print (L"Could not allocate memory for Token.\n");
+      return EFI_OUT_OF_RESOURCES;
+    }
+  }  ///< All Arguments analysed
+
+  if (ParameterCount == 0) {
+    Print (L"No option selected, verify only.\n");
+    Print (L"Usage: ControlMsrE2 <unlock | lock | interactive>\n\n");
+  } else if (ParameterCount > 1) {
+    Print (L"interactive, unlock, lock are exclusive options. Use only one of them.\n\n");
+    return EFI_INVALID_PARAMETER;
+  }
+
+  return EFI_SUCCESS;
+}
+
+VOID
+ModifySearchString (
+  IN OUT EFI_STRING *SearchString
+  )
+{
+  BOOLEAN   Result;
+
+  do {
+    Print (L"\nCurrent search string: %s\n", *SearchString);
+    Print (L"Do you want to change it ? ");
+    Result = ReadYN ();
+    if (Result) {
+      Print (L"\nEnter search string: ");
+
+      CHAR16 *Buffer = AllocatePool (BUFFER_LENGTH * sizeof (CHAR16));
```

Comment:
Might as well use the stack if you opt for a relatively small constant-size buffer, provided by the caller (would remove the allocation from it as well).